diff --git a/bin/js-yaml.js b/bin/js-yaml.js
--- a/bin/js-yaml.js
+++ b/bin/js-yaml.js
@@ -23,42 +23,46 @@
 
 var cli = new argparse.ArgumentParser({
   prog:     'js-yaml',
-  version:  require('../package.json').version,
-  addHelp:  true
+  add_help:  true
 });
 
 
-cli.addArgument([ '-c', '--compact' ], {
+cli.add_argument('-v', '--version', {
+  action: 'version',
+  version: require('../package.json').version
+});
+
+cli.add_argument('-c', '--compact', {
   help:   'Display errors in compact mode',
-  action: 'storeTrue'
+  action: 'store_true'
 });
 
 
 // deprecated (not needed after we removed output colors)
 // option suppressed, but not completely removed for compatibility
-cli.addArgument([ '-j', '--to-json' ], {
-  help:   argparse.Const.SUPPRESS,
+cli.add_argument('-j', '--to-json', {
+  help:   argparse.SUPPRESS,
   dest:   'json',
-  action: 'storeTrue'
+  action: 'store_true'
 });
 
 
-cli.addArgument([ '-t', '--trace' ], {
+cli.add_argument('-t', '--trace', {
   help:   'Show stack trace on error',
-  action: 'storeTrue'
+  action: 'store_true'
 });
 
-cli.addArgument([ 'file' ], {
+cli.add_argument('file', {
   help:   'File to read, utf-8 encoded without BOM',
   nargs:  '?',
-  defaultValue: '-'
+  default: '-'
 });
 
 
 ////////////////////////////////////////////////////////////////////////////////
 
 
-var options = cli.parseArgs();
+var options = cli.parse_args();
 
 
 ////////////////////////////////////////////////////////////////////////////////
